### ISSUE TITLE
Shared config doesn't cascade properly

### ISSUE THREAD

**Issue by dominicbarnes** [eslint/eslint#2592]
I've been trying to isolate this bug, but it seems to only occur under very specific conditions. I'll continue trying to boil this down, but for now I'll report what I know.

I have a company shared-config, with 2 modes: `node` and `browser` Our app's config hierarchy looks like:

 * `.eslintrc`: `{ extends: "segment/node" }`
 * `app/client/.eslintrc`: `{ extends: "segment/browser" }`

When ESLint runs, it approaches files in this order:

 * `app/client/index.js`
 * `app/server.js`

What happens is that the `server.js` file (which should _only_ be following `segment/node`) is wrongly failing because of rules in `segment/browser`. If I manually reverse the order, it processes the rules correctly. (which tells me that some clobbering is happening between various config objects)

I've dropped a couple `console.log` in there to try and follow the cascade and see where things go awry. The first time `segment/node` is loaded, it appears to get the right value, but all subsequent times get the wrong value.

From what I can see, `utils.mergeConfig()` modifies the `base` directly, which I suspect is the source of this problem. I'm hoping the solution can be as simple as cloning the config objects after being `require`d as to prevent clobbering.

**Comment by gyandeeps** [comment]
What version of eslint are you running?

**Comment by dominicbarnes** [comment]
It is the latest version, `0.21.2`

**Comment by gyandeeps** [comment]
based on how you have solved this problem in your PR, it tells me that `assign` module does a shallow copy and not a deep copy.
Because if it does a deep copy then this problem should not come up.
I am saying this based on how you have solved the problem as I still have to look at the code at runtime.

**Comment by dominicbarnes** [comment]
Gotcha, yeah `object-assign` is definitely a shallow copy.

**Comment by ilyavolodin** [comment]
@gyandeeps You are correct. `object-assign` is only shallow copy.

**Comment by gyandeeps** [comment]
@ilyavolodin I agree. If we have a way to do a deep copy without including a new dependency then it awesome.

**Comment by dominicbarnes** [comment]
Do you guys mean something like:

```js
util.mergeConfigs({}, require(filePath));
```

instead of:

```js
clone(require(filePath));
```

**Comment by dominicbarnes** [comment]
I've updated my PR accordingly, it appears to solve the problem equally well :)

**Comment by ilyavolodin** [comment]
To be more specific
```js
var newConfig = util.mergeConfigs({}, require(filePath));
```
But yes, mergeConfigs was created to do deep copy of config objects.

**Comment by nzakas** [comment]
@ilyavolodin @gyandeeps have either of you confirmed that this is a bug?

**Comment by ilyavolodin** [comment]
I haven't tried it myself, but looking at the code, and knowing that @gyandeeps confirmed it last time, this sounds about right.

**Comment by nzakas** [comment]
Ok, marking as a bug
